stdlib/public/core/StringGuts.swift

```diff
@@ -0,0 +1,1355 @@
+// TODO
+import SwiftShims
+
+//
+// Largely inspired by string-recore at
+// https://github.com/apple/swift/pull/10747.
+//
+
+//
+// StringGuts is always 16 bytes on both 32bit and 64bit platforms. This
+// effectively makes the worse-case (from a spare bit perspective) ABIs be the
+// 64bit ones, as all 32bit pointers effectively have a 32-bit address space
+// while the 64bit ones have a 56-bit address space.
+//
+// Of the 64bit ABIs, x86_64 has the fewest spare bits, so that's the ABI we
+// design for.
+//
+// FIXME: what about ppc64 and s390x?
+//
+
+
+
+@_fixed_layout
+public // FIXME
+struct _StringGuts {
+  // TODO 32-bit: Insert padding between fields
+  public // FIXME for testing only
+  var _storage: (_StringObject, UInt)
+
+  @_inlineable
+  public
+  var _object: _StringObject {
+    @inline(__always) get { return _storage.0 }
+    @inline(__always) set { _storage.0 = newValue }
+  }
+  @_inlineable
+  public
+  var _otherBits: UInt {
+    @inline(__always) get { return _storage.1 }
+    @inline(__always) set { _storage.1 = newValue }
+  }
+
+  @_inlineable
+  @inline(__always)
+  public
+  init(object: _StringObject, otherBits: UInt) {
+    self._storage.0 = object
+    self._storage.1 = otherBits
+    _invariantCheck()
+  }
+}
+
+extension _StringGuts {
+  @_inlineable // FIXME(sil-serialize-all)
+  @_versioned // FIXME(sil-serialize-all)
+  internal func _invariantCheck() {
+#if INTERNAL_CHECKS_ENABLED
+    if _object.isNative {
+      _sanityCheck(UInt(_object.nativeRawStorage.count) == self._otherBits)
+    } else if _object.isUnmanaged {
+    } else if _object.isCocoa {
+      if _object.isContiguous {
+        _sanityCheck(_otherBits != 0) // TODO: in ABI's address space
+      } else {
+        _sanityCheck(_otherBits == 0)
+      }
+    } else if _object.isSmall {
+    } else {
+      fatalError("Unimplemented string form")
+    }
+#endif
+  }
+
+  @_inlineable
+  @inline(__always)
+  public // @testable
+  mutating func isUniqueNative() -> Bool {
+    guard _isNative else { return false }
+    // Note that the isUnique test must be in a separate statement;
+    // `isNative && _isUnique` always evaluates to false in debug builds,
+    // because SILGen keeps the self reference in `isNative` alive for the
+    // duration of the expression.
+
+    // Note that we have to perform this operation here, and not as a (even
+    // mutating) method on our _StringObject to avoid all chances of a semantic
+    // copy.
+    //
+    // FIXME: Super hacky. Is there a better way?
+    defer { _fixLifetime(self) }
+    var bitPattern = _object.payloadBits
+    return _isUnique_native(&bitPattern)
+  }
```

Comment:
CC @atrick, this is what we were discussing earlier.

Reply:
I have a better idea now how SILGen could be fixed to allow isUnique(&self.property). I'm not sure when someone would get around to fixing it, but filing a bug for any case you come across would be reasonable and actionable.